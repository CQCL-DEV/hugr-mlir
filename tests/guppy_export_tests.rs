
use rstest::{rstest,fixture};

use hugr_mlir::Result;

#[fixture]
pub fn test_context() -> melior::Context {
    let dr = melior::dialect::DialectRegistry::new();
    let hugr_dh = hugr_mlir::mlir::hugr::get_hugr_dialect_handle();
    hugr_dh.insert_dialect(&dr);
    let ctx = melior::Context::new();
    ctx.append_dialect_registry(&dr);
    ctx.load_all_available_dialects();
    ctx
}

#[rstest]
fn test_guppy_exports(test_context: melior::Context) -> Result<()> {
    insta::glob!("guppy-exports", "*.json", |path| {
        use std::fs;
        let src = fs::read(path).unwrap();
        let ul = melior::ir::Location::new(&test_context, path.to_str().unwrap(), 0, 0);
        let m = hugr_mlir::translate::translate_hugr_to_mlir(&src, ul, |bytes| serde_json::from_slice::<hugr::Hugr>(bytes)).unwrap();
        let o = m.as_operation();
        println!("{o}");
        assert!(m.as_operation().verify());
        insta::assert_display_snapshot!(m.as_operation());
    });
    Ok(())
}
