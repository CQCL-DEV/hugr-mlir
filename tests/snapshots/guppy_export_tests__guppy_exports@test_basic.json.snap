---
source: tests/guppy_export_tests.rs
expression: m.as_operation()
input_file: tests/guppy-exports/test_basic.json
---
module {
  hugr.module {
    func @foo[](%arg0: !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> {
      %0 = cfg %arg0 : (!hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> {
      ^bb0(%arg1: !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>):
        func @bar[](%arg2: !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> {
          %8 = cfg %arg2 : (!hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> {
          ^bb0(%arg3: !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>):
            %9 = make_tuple()
            %10 = tag 0 %9 : tuple<> -> <tuple<>>
            switch %10, %arg3 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> ^bb1
          ^bb1(%11: !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>):  // pred: ^bb0
            output %11 : !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>
          }
          output %8 : !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>
        }
        %1 = load_constant @bar : !hugr<function [](!hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>>
        const @const_20 : !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> = 1 : i6
        %2 = load_constant @const_20 : !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>
        %3 = ext_op []"iadd" %arg1, %2 : (!hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>, !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>
        %4 = call %1  [](!hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>) -> !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> %3
        %5 = make_tuple()
        %6 = tag 0 %5 : tuple<> -> <tuple<>>
        switch %6, %4 : !hugr.sum<tuple<>>, !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>> ^bb1
      ^bb1(%7: !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>):  // pred: ^bb0
        output %7 : !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>
      }
      output %0 : !hugr<opaque "int"["arithmetic.int.types"] [Copyable] <6 : i64>>
    }
  }
}

